edi_storage/models/edi_backend.py

```diff
@@ -76,3 +82,79 @@ def _component_sort_key(self, component_class):
+
+    def _storage_check_pending_input(self, **kw):
+        """Create new exchange records if new files found.
+
+        Collect input exchange types and for each of them,
+        check by pattern if the a new exchange record is required.
+        """
+        self.ensure_one()
+        if not self.storage_id or not self.input_dir_pending:
+            _logger.info(
+                "%s ignored: no storage and/or input directory specified.", self.name
+            )
+            return False
+
+        exchange_types = self.env["edi.exchange.type"].search(
+            self._storage_exchange_type_pending_input_domain()
+        )
+        for exchange_type in exchange_types:
+            # NOTE: this call might keep hanging the cron
+            # if the remote storage is slow (eg: too many files)
+            # We should probably run this code in a separate job per exchange type.
+            file_names = self._storage_get_input_filenames(exchange_type)
+            _logger.info(
+                "Processing exchange type '%s': found %s files to process",
+                exchange_type.display_name,
+                len(file_names),
+            )
+            for file_name in file_names:
+                # TODO: add config for job
+                self.with_delay()._storage_create_record_if_missing(
+                    exchange_type, file_name
+                )
+        return True
+
+    def _storage_exchange_type_pending_input_domain(self):
+        """Domain for retrieving input exchange types.
+        """
+        return [
+            ("backend_type_id", "=", self.backend_type_id.id),
+            ("direction", "=", "input"),
+        ]
+
+    def _storage_create_record_if_missing(self, exchange_type, remote_file_name):
+        """Create a new exchange record for given type and file name if missing.
+        """
+        file_name = os.path.basename(remote_file_name)
+        extra_domain = [("exchange_filename", "=", file_name)]
+        existing = self._find_existing_exchange_records(
+            exchange_type, extra_domain=extra_domain, count_only=True
+        )
+        if existing:
+            return
+        record = self.create_record(
+            exchange_type.code, self._storage_new_exchange_record_vals(file_name)
+        )
+        _logger.debug("%s: new exchange record generated.", self.name)
+        return record.identifier
+
+    def _storage_get_input_filenames(self, exchange_type):
+        if not exchange_type.exchange_filename_pattern:
+            # If there is not patter, return everything
+            return self.storage_id.list_files(self.input_dir_pending)
+
+        # TODO: validate pattern usage.
+        bits = [exchange_type.exchange_filename_pattern]
+        if exchange_type.exchange_file_ext:
+            bits.append("*." + exchange_type.exchange_file_ext)
+        pattern = "".join(bits)
+        return self.storage_id.find_files(pattern, self.input_dir_pending)
```

Comment:
@simahawk  I think, we cannot really use the pattern here. To convert the pattern we need a exchange records (see https://github.com/OCA/edi/blob/13.0/edi/models/edi_exchange_type.py#L134), but here were about to create it so we don't have one.

The maximun we could do is something like this:
```
pattern.format(
            exchange_record="",
            record="",
            record_name="",
            type=exchange_type,
            dt=dt,
        )
```
 What do you think? is there anything I'm missing?

Reply:
This is my proposal: https://github.com/OCA/edi/pull/402/commits/4cda6390d9cbf36d05506b1ba3cd8857b8843967. Let me know if you see possible improvements :)

Reply:
> @simahawk I think, we cannot really use the pattern here. To convert the pattern we need a exchange record (see https://github.com/OCA/edi/blob/13.0/edi/models/edi_exchange_type.py#L134), but here were about to create it so we don't have one.
> 
> The maximun we could do is something like this:

I don't think that's completely correct. The pattern is in your control. You decide if you need the exchange record or not.

And in this case, you don't care about `exchange_filename` name because you are going to set it when creating the record (it's computed but write-able).

So, what you put in the pattern does not have to satisfy `_make_exchange_filename` because that method is called only if the filename is not set.

I'd rollback this change, unless I'm missing something else.

Reply:
So, then what do you expect users to enter in the pattern field? a regex? 

To make things more clear and explicit, confirm if the following is correct: your idea is to use the `exchange_filename_pattern` in two different ways:

- to generate the exchange filename in and output type. pattern is to use available variables between brackets, e.g: `{exchange_record}-{type}-{dt}`.
- to find/filter files to be imported from an storage for the input exchange type. E.g: `some-type.*.xml`

Please, correct me if I'm misunderstanding your proposal.

Reply:
This! :smiley: 
We could add an help text in the form maybe when edi_storage is installed

Reply:
Awesome, done in https://github.com/OCA/edi/pull/402/commits/90e00436b1dc57682f3617904d30084f4fd3b59b.